### PR TITLE
Add a permissions system to restrict E2 functions

### PR DESCRIPTION
**Pull request by garryspins**

### CODE REVIEW

File: lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua [hunk 1]
```diff
@@ -1,8 +1,15 @@
-
+--[[
+	Expression 2 Permissions
+	
+]]
 AddCSLuaFile()
 
-E2Permissions = E2Permissions or {}
+PIXEL = PIXEL or {}
+PIXEL.E2Permissions = {}
```
Comment (TomDotBat) [inline lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua:10]: Add auto refresh support, define PIXEL.E2Permissions like this:
```lua
PIXEL.E2Permissions = PIXEL.E2Permissions or {
    Whitelist = {},
    SuperWhitelist = {},
    Blacklist = {},
    WhitelistedFunctions = {},
    BlacklistedFunctions = {}
}
```

File: lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua [hunk 2]
```diff
@@ -3,14 +3,33 @@
...
+
+local w, sw, b, wf, bf = PIXEL.E2Permissions.Whitelist, PIXEL.E2Permissions.SuperWhitelist, PIXEL.E2Permissions.Blacklist, PIXEL.E2Permissions.WhitelistedFunctions, PIXEL.E2Permissions.BlacklistedFunctions
+local t, f = true, false
 
 function PIXEL.E2Permissions.CanInvoke(ply, name, params, meta)
-    return not not hook.Run("PIXEL.E2Invoke", ply, name, params, meta) -- the !! is for boolean conversion, its just nicer
+    return hook.Run("PIXEL.E2Permissions.CanInvoke", ply, name, params, meta)
 end
+
+hook.Add("PIXEL.E2Permissions.CanInvoke","MainInvokeStop", function(ply, name, params, meta)
```
Comment (TomDotBat) [inline lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua:24]: Space out arguments correctly and localise all functions called and tables indexed in this hook.

File: lua/wire/client/text_editor/wire_expression2_editor.lua [hunk 3]
```diff
@@ -1805,7 +1805,7 @@ function Editor:LoadFile(Line, forcenewtab)
 
 	local f = file.Open(Line, "r", "DATA")
 	if not f then
-		ErrorNoHalt("Erroring opening file: " .. Line)
+		-- ErrorNoHalt("Erroring opening file: " .. Line)
```
Comment (TomDotBat) [inline lua/wire/client/text_editor/wire_expression2_editor.lua:1808]: Why was this line removed? This could can cause errors further along as you're not even returning. At least fix that and provide some sort of feedback.

File: lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua [hunk 4]
```diff
@@ -3,14 +3,33 @@
...
-PIXEL.E2Permissions.Whitelist = {}
-PIXEL.E2Permissions.Blacklist = {}
-PIXEL.E2Permissions.WhitelistOnlyFunctions = {}
+PIXEL.E2Permissions.Whitelist = {} 				-- Access to normal functions and whitelisted functions
+PIXEL.E2Permissions.SuperWhitelist = {} 		-- Access to EVERY function
+PIXEL.E2Permissions.Blacklist = {}				-- Cannot access any E2 function
+PIXEL.E2Permissions.WhitelistedFunctions = {}	-- Functions Whitelisted people can access
+PIXEL.E2Permissions.BlacklistedFunctions = {} 	-- Functions Noone Except SuperWhitelisted Users can Access
+
+local w, sw, b, wf, bf = PIXEL.E2Permissions.Whitelist, PIXEL.E2Permissions.SuperWhitelist, PIXEL.E2Permissions.Blacklist, PIXEL.E2Permissions.WhitelistedFunctions, PIXEL.E2Permissions.BlacklistedFunctions
```
Comment (TomDotBat) [inline lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua:17]: What's with all the unused localised variables? If you do use them, make their names more descriptive first.

File: lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua [hunk 5]
```diff
@@ -3,14 +3,33 @@
...
 
 function PIXEL.E2Permissions.CanInvoke(ply, name, params, meta)
-    return not not hook.Run("PIXEL.E2Invoke", ply, name, params, meta) -- the !! is for boolean conversion, its just nicer
+    return hook.Run("PIXEL.E2Permissions.CanInvoke", ply, name, params, meta)
 end
+
+hook.Add("PIXEL.E2Permissions.CanInvoke","MainInvokeStop", function(ply, name, params, meta)
+	if not IsValid(ply) then return false end
+	if PIXEL.E2Permissions.SuperWhitelist[ply:SteamID64()] then return true end
+	if PIXEL.E2Permissions.Blacklist[ply:SteamID64()] then return true end
```
Comment (TomDotBat) [inline lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua:27]: Store the player's SteamID64 in a local here instead of potentially getting it multiple times.

**Review by TomDotBat — CHANGES_REQUESTED** [review]

File: lua/wire/client/text_editor/wire_expression2_editor.lua (hunk 3 above)
Comment (garryspins) [reply lua/wire/client/text_editor/wire_expression2_editor.lua:1808]: This is something i have to look over, forgot about it but it was causing errors for no logical reason

Comment (TomDotBat) [reply lua/wire/client/text_editor/wire_expression2_editor.lua:1808]: Possibly broken by Lua refresh? I have't seen this happen anywhere else before.

Comment (garryspins) [reply lua/wire/client/text_editor/wire_expression2_editor.lua:1808]: probably

File: lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua (hunk 4 above)
Comment (garryspins) [reply lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua:17]: Dealt with

File: lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua (hunk 2 above)
Comment (garryspins) [reply lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua:24]: Done

File: lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua (hunk 1 above)
Comment (garryspins) [reply lua/entities/gmod_wire_expression2/permissions/e2_permissions.lua:10]: Done

File: lua/entities/gmod_wire_expression2/base/compiler.lua [hunk 6]
```diff
@@ -22,6 +22,7 @@ function Compiler:Error(message, instr)
 end
 
 function Compiler:Process(root, inputs, outputs, persist, delta, includes, ply) -- Took params out becuase it isnt used.
+	local canInvoke = PIXEL.E2Permissions.CanInvoke
```
Comment (TomDotBat) [inline lua/entities/gmod_wire_expression2/base/compiler.lua:25]: Localise `CanInvoke` outside of this method

**Review by TomDotBat — CHANGES_REQUESTED** [review]

**Review by TomDotBat — APPROVED** [review]
